### PR TITLE
Verify header field names

### PR DESCRIPTION
**Pull request by fabianfett**
This pr builds on top of #189 and will need a rebase once the former is merged.

### Motivation

Having spent about four hours today debugging AsyncHTTPClient just to find out I had a whitespace instead of a dash in my http header field name, I decided I want to save somebody else's time – probably future me.

`User Agent` vs `User-Agent`

### Changes

- Added a method: `HTTPHeaders.validateFieldNames() throws`
- Added an error `.invalidHeaderFieldNames([String])`

### Result

Hours of time saved. Karma for me.

### CODE REVIEW

File: Sources/AsyncHTTPClient/RequestValidation.swift [hunk 1]
```diff
@@ -29,28 +29,35 @@ extension HTTPHeaders {
...
-                throw HTTPClientError.chunkedSpecifiedMultipleTimes
-            }
+        try self.validateFieldNames()
+
+        guard let body = body else {
+            // https://tools.ietf.org/html/rfc2616#section-4.3
+            //
+            // The presence of a message-body in a request is signaled by the inclusion of a
+            // Content-Length or Transfer-Encoding header field in the request's message-headers.
+            return
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/RequestValidation.swift]: RFC 2616 is dead, let's stop citing it. It's RFC 7230 section 3.3. _However_, this change is not cautious enough: if there's no body, but the request verb is `POST`, "Content-Length: 0" is mandatory. Content-Length: 0 is also harmless on all request types that do not define a body. A patch that users this escape hatch on only `GET`, `HEAD`, `DELETE`, `CONNECT`, and `TRACE` would be acceptable though, so long as it sends C-L 0 everywhere else.

**Comment by fabianfett** [comment]
@Lukasa Compile errors are fixed. There are some failing tests though (nothing I touched).

**Comment by artemredkin** [comment]
@fabianfett it seems that most tests fail due to change in behaviour, `Content-Length: 0` was expected in those tests, but now it's missing

**Comment by fabianfett** [comment]
@artemredkin ow shoot. fixed.

**Comment by artemredkin** [comment]
awesome, thanks! there is one failing test, but this one is flaky (#175), I'll try to run tests again

**Comment by artemredkin** [comment]
@swift-server-bot test this please

File: Tests/AsyncHTTPClientTests/RequestValidationTests.swift [hunk 2]
```diff
@@ -0,0 +1,95 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
```
Comment (weissi) [inline Tests/AsyncHTTPClientTests/RequestValidationTests.swift:5]: @fabianfett can we fix the copyright here? should be `2020` because it's all new code

**Comment by weissi** [comment]
@fabianfett From what I can tell, we're good but would you mind checking if we have a test case which covers this https://github.com/swift-server/async-http-client/issues/146 ? If yes, we're all good, if no we might want to add it here?

File: Tests/AsyncHTTPClientTests/RequestValidationTests.swift [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            ("Transfer-Encoding", "chunked"),
+        ])
+        var buffer = ByteBufferAllocator().buffer(capacity: 200)
+        buffer.writeBytes([UInt8](repeating: 12, count: 200))
+        XCTAssertThrowsError(try headers.validate(method: .TRACE, body: .byteBuffer(buffer))) {
+            XCTAssertEqual($0 as? HTTPClientError, .traceRequestWithBody)
+        }
+    }
+
+    func testGET_HEAD_DELETE_CONNECTRequestCanHaveBody() {
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/RequestValidationTests.swift:74]: @weissi this is for you.

File: Tests/AsyncHTTPClientTests/RequestValidationTests.swift (hunk 2 above)
Comment (fabianfett) [reply Tests/AsyncHTTPClientTests/RequestValidationTests.swift]: fixed

**Comment by artemredkin** [comment]
@swift-server-bot test this please

**Comment by artemredkin** [comment]
@swift-server-bot test this please

**Comment by artemredkin** [comment]
@swift-server-bot test this please

File: Tests/AsyncHTTPClientTests/RequestValidationTests.swift (hunk 3 above)
Comment (weissi) [reply Tests/AsyncHTTPClientTests/RequestValidationTests.swift:74]: thanks so much!

**Comment by artemredkin** [comment]
thank you @fabianfett !

**Pull request merged**